### PR TITLE
feat: Upgrade to DataFusion 40.0.0-rc1

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

DataFusion is preparing to release version 40.0.0 so I would like to make sure that we are testing with the latest prior to the release.

## What changes are included in this PR?

Use a very recent revision of the official DataFusion repo.

### CODE REVIEW

**Comment by andygrove** [comment]
There is a test failure:

```
- Windows support *** FAILED *** (224 milliseconds)
  org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1241.0 failed 1 times, most recent failure: Lost task 0.0 in stage 1241.0 (TID 4084) (192.168.64.28 executor driver): org.apache.comet.CometNativeException: General execution error with reason: count not supported for window function.
  ```

**Comment by viirya** [comment]
> There is a test failure:
> 
> ```
> - Windows support *** FAILED *** (224 milliseconds)
>   org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1241.0 failed 1 times, most recent failure: Lost task 0.0 in stage 1241.0 (TID 4084) (192.168.64.28 executor driver): org.apache.comet.CometNativeException: General execution error with reason: count not supported for window function.
> ```

I recall that count, max, min are supported as window function recently. Maybe there are some changes in DataFusion.

cc @huaxingao

**Comment by huaxingao** [comment]
> I recall that count, max, min are supported as window function recently. Maybe there are some changes in DataFusion.

maybe this [PR](https://github.com/apache/datafusion/pull/10484)? I need to take a closer look.

**Comment by andygrove** [comment]
> > I recall that count, max, min are supported as window function recently. Maybe there are some changes in DataFusion.
> 
> maybe this [PR](https://github.com/apache/datafusion/pull/10484)? I need to take a closer look.

We rely on `find_df_window_func` to find window functions, and `count` was removed from the built-in functions in https://github.com/apache/datafusion/pull/10893

**Comment by andygrove** [comment]
@viirya @huaxingao I made a change so that we fall back to Spark for count in window aggregate for now, and I will file a follow on issue. This PR is ready for review now.

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 1]
```diff
@@ -208,7 +208,7 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
       expr match {
         case agg: AggregateExpression =>
           agg.aggregateFunction match {
-            case _: Min | _: Max | _: Count =>
```
Comment (viirya) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:211]: This disables `count` for both aggregation and window?

Comment (viirya) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:211]: Oh, nvm, I checked the source code. This is only for window functions.

Comment (viirya) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:211]: Can we add a comment here?

**Review by viirya — APPROVED** [review]

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 1 above)
Comment (andygrove) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:211]: Added

**Review by kazuyukitanimura — APPROVED** [review]

**Pull request merged**